cleanlab/data_valuation.py

```diff
@@ -0,0 +1,134 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+"""
+Provides methods for computing the Shapley values of data points using a K-Nearest Neighbors (KNN) graph.
+This approach allows for the assessment of individual data points' contributions to the model's performance in a dataset.
+The module's primary function, `data_shapley_knn`, enables the computation of Shapley values either directly from data features or using a precomputed KNN graph.
+Suitable for scenarios where understanding the significance of each data point in model training is crucial.
+
+The methodology is based on the approach described in https://arxiv.org/abs/1911.07128.
+The paper discusses balancing scalability and utility in data importance, suggesting KNN-based Shapley value approximation as a viable approach.
+The implementation here seeks to make these insights accessible for practical usage, offering an efficient way to quantify data point importance in ML workflows.
+"""
+
+
+from typing import Optional, cast
+
+import numpy as np
+from scipy.sparse import csr_matrix
+from sklearn.neighbors import NearestNeighbors
+from sklearn.exceptions import NotFittedError
+from sklearn.utils.validation import check_is_fitted
+
+
+def _knn_shapley_score(knn_graph: csr_matrix, labels: np.ndarray, k: int) -> np.ndarray:
+    """Compute the Shapley values of data points based on a knn graph."""
+    N = labels.shape[0]
+    scores = np.zeros((N, N))
+    dist = knn_graph.indices.reshape(N, -1)
+
+    for y, s, dist_i in zip(labels, scores, dist):
+        idx = dist_i[::-1]
+        ans = labels[idx]
+        s[idx[k - 1]] = float(ans[k - 1] == y)
+        ans_matches = (ans == y).flatten()
+        for j in range(k - 2, -1, -1):
+            s[idx[j]] = s[idx[j + 1]] + float(int(ans_matches[j]) - int(ans_matches[j + 1]))
+    return 0.5 * (np.mean(scores / k, axis=0) + 1)
+
+
+def _process_knn_graph_from_features(
+    features: np.ndarray, metric: Optional[str], k: int = 10
+) -> csr_matrix:
+    """Calculate the knn graph from the features if it is not provided in the kwargs."""
+    if k > len(features):  # Ensure number of neighbors less than number of examples
+        raise ValueError(
+            f"Number of nearest neighbors k={k} cannot exceed the number of examples N={len(features)} passed into the estimator (knn)."
+        )
+    if metric == None:
+        metric = "cosine" if features.shape[1] > 3 else "euclidean"
+    knn = NearestNeighbors(n_neighbors=k, metric=metric).fit(features)
+    knn_graph = knn.kneighbors_graph(mode="distance")
+    try:
+        check_is_fitted(knn)
+    except NotFittedError:
+        knn.fit(features)
+    return knn_graph
+
+
+def data_shapley_knn(
+    labels: np.ndarray,
```

Comment:
```suggestion
    labels: np.ndarray, *,
```

Reply:
we always want to be able to re-order the keyword only args for every method in cleanlab package